src/engine/sync/enginesync.cpp

```diff
@@ -19,14 +19,200 @@ EngineSync::EngineSync(UserSettingsPointer pConfig)
+    // decks that need to change as a result.
+    Syncable* oldMaster = nullptr;
+    switch (mode) {
+    case SYNC_MASTER_EXPLICIT: {
+        if (pSyncable->getBaseBpm() > 0 || pSyncable == m_pInternalClock) {
+            activateMaster(pSyncable, SYNC_MASTER_EXPLICIT);
+        } else {
+            // If we have no bpm, we can't be a valid master. Override.
+            activateFollower(pSyncable);
+        }
+        break;
+    }
+    case SYNC_MASTER_SOFT: {
+        if (m_pMasterSyncable == pSyncable &&
+                pSyncable->getSyncMode() == SYNC_MASTER_EXPLICIT) {
+            pSyncable->setSyncMode(mode);
+        } else if (pSyncable->getBaseBpm() > 0.0) {
+            activateMaster(pSyncable, SYNC_MASTER_SOFT);
+        } else {
+            activateFollower(pSyncable);
+        }
+        break;
+    }
+    case SYNC_FOLLOWER: {
+        // A request for follower mode may be converted into an enabling of soft
+        // master mode.
+        if (m_pMasterSyncable == pSyncable) {
+            m_pMasterSyncable = nullptr;
+            activateFollower(pSyncable);
+        } else {
+            Syncable* newMaster = pickMaster(pSyncable);
+            if (newMaster && newMaster != m_pMasterSyncable) {
+                // if the master has changed, activate it (this updates m_pMasterSyncable)
+                activateMaster(newMaster, SYNC_MASTER_SOFT);
+            }
+            if (newMaster != pSyncable) {
+                activateFollower(pSyncable);
+            }
+        }
+        break;
+    }
+    case SYNC_NONE: {
+        if (pSyncable != m_pInternalClock) {
+            deactivateSync(pSyncable);
+        }
+        break;
+    }
+    default:;
+    }
+
+    // Second, figure out what Syncable should be used to initialize the master
+    // parameters, if any. Usually this is the new master. (Note, that pointer might be null!)
+    Syncable* pParamsSyncable = m_pMasterSyncable;
+    // But If we are newly soft master, we need to match to some other deck.
+    if (pSyncable == m_pMasterSyncable && pSyncable != oldMaster && mode != SYNC_MASTER_EXPLICIT) {
+        pParamsSyncable = findBpmMatchTarget(pSyncable);
+        if (!pParamsSyncable) {
+            // We weren't able to find anything to match to, so set ourselves as the
+            // target.  That way we'll use our own params when we setMasterParams below.
+            pParamsSyncable = pSyncable;
+        }
+    }
+    // Now that all of the decks have their assignments, reinit master params if needed.
+    if (pParamsSyncable) {
+        setMasterParams(pParamsSyncable);
+        pSyncable->setInstantaneousBpm(pParamsSyncable->getBpm());
+        if (pParamsSyncable != pSyncable) {
+            pSyncable->requestSync();
+        }
+    }
+
+    if (auto unique_syncable = getUniquePlayingSyncable(); unique_syncable) {
+        unique_syncable->notifyOnlyPlayingSyncable();
+    }
```

Comment:
```suggestion
    const auto* uniqueSyncable = getUniquePlayingSyncable();
    if (uniqueSyncable != nullptr) {
        uniqueSyncable->notifyOnlyPlayingSyncable();
    }
```

Reply:
we're not doing c++17 conditionals? but they are so clean!

Reply:
I agree that doing assignment inside conditional expressions is a bad practice. It is IMO an misfeature of the language that it compiles.

Reply:
It compiles because it's a specifically-added new feature: https://skebanga.github.io/if-with-initializer/.  As for the danger of assignments in conditionals, the compiler should throw a warning for assignments without associated checks But I'll remove it.

Reply:
(the pointer can't be const because the function is non-const)

Reply:
But the lvalue can be const, right?

Reply:
it didn't compile, so shrug

Reply:
IIRC for plain pointers the const needs to go behind pointer type. But doesn't Metter, let's just leave it.